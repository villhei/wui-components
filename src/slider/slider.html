<style>
  input {
    -webkit-appearance: none;
    flex: 1;
    height: 15px;
    border-radius: 5px;
    background: #d3d3d3;
    outline: none;
    opacity: 0.7;
    -webkit-transition: 0.2s;
    transition: opacity 0.2s;
  }

  input::-webkit-slider-thumb {
    -webkit-appearance: none;
    appearance: none;
  }

  input::-webkit-slider-thumb {
    width: 25px;
    height: 25px;
    border-radius: 50%;
    background: #04aa6d;
    cursor: pointer;
  }

  input::-moz-range-thumb {
    width: 25px;
    height: 25px;
    border-radius: 50%;
    background: #04aa6d;
    cursor: pointer;
  }

  :host {
    display: inline-flex;
    flex-direction: column;
    padding: var(--component-padding);
  }
</style>
<input type="range" min="1" max="100" value="50" class="slider" id="myRange" />
