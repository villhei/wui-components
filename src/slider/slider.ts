import templateContent from "./slider.html?raw";
import { WUIBase } from "../core";

const template = document.createElement("template");

template.innerHTML = templateContent;

const variants = ["success", "warning", "error"];

export class Slider extends WUIBase {
  static formAssociated = true;

  constructor() {
    super(template);
    this.inputNode = this.root.querySelector("input")!;
    this.labelNode = this.root.querySelector("label")!;
    this.messageNode = this.root.querySelector("#message")!;
    this.internals = this.attachInternals();
    this.update();
    this._value = "";
  }

  private _value: string;
  internals: ElementInternals;
  inputNode: HTMLInputElement;
  labelNode: HTMLLabelElement;
  messageNode: HTMLElement;

  static get observedAttributes() {
    return ["class", "label", "name", "disabled", "value", "type", "message"];
  }

  handleInputEvent = (event: Event) => {
    const clone = new Event(event.type, event);
    this.value = this.inputNode.value;

    this.dispatchEvent(clone);
  };

  connectedCallback() {
    const initialValue = this.getAttribute("value");
    if (initialValue !== null) {
      this.inputNode.setAttribute("value", initialValue);
    }
    this.inputNode.addEventListener("change", this.handleInputEvent);
  }

  disconnectedCallback() {
    this.inputNode.removeEventListener("change", this.handleInputEvent);
  }

  get label() {
    return this.getAttribute("label");
  }

  get name() {
    return this.getAttribute("name");
  }

  get disabled() {
    return this.getBoolean("disabled");
  }

  get placeholder() {
    return this.getAttribute("placeholder");
  }

  get classes() {
    return this.getAttribute("class")?.split(",") || [];
  }

  set value(value) {
    this._value = value;
    this.inputNode.value = value;
    if (this.internals) {
      this.internals.setFormValue(value);
    }
  }

  get value() {
    return this._value;
  }

  get type() {
    return this.getAttribute("type") || "text";
  }

  get variant() {
    return this.getAttribute("variant") || "default";
  }

  get message() {
    return this.getAttribute("message");
  }

  update() {
    this.classes.forEach((className) =>
      this.inputNode.classList.add(className)
    );
    this.labelNode.textContent = this.label;
    if (this.name) {
      this.labelNode.setAttribute("for", this.name);
      this.inputNode.setAttribute("name", this.name);
    }
    this.forwardAttribute("placeholder", this.inputNode);
    this.forwardAttribute("disabled", this.inputNode);
    this.forwardAttribute("type", this.inputNode);
    this.forwardAttribute("variant", this.inputNode);

    if (this.message) {
      this.messageNode.textContent = this.message;
    } else {
      this.messageNode.textContent = "";
    }
  }
}
