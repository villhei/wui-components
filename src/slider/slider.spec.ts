import { Slider } from "./slider";
import { fireEvent } from "@testing-library/dom";

const attachInternalsMock = jest.fn(() => {
  return {
    setFormValue: jest.fn(),
  };
});
beforeAll(() => {
  customElements.define("wui-slider", Slider);
});

beforeEach(() => {
  (Element.prototype as any).attachInternals = attachInternalsMock;
});

afterEach(() => {
  jest.resetAllMocks();
});

it("should match snapshot", () => {
  const el = new Slider();
  expect(el.shadowRoot?.innerHTML).toMatchSnapshot();
});
