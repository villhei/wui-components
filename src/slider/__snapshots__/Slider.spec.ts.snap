// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
"<style>
  :host {
    --input-focused-border-color: var(--primary-color);
    --font-size: var(--font-size);
    --component-padding: 0.5rem;
    --inactive-background: var(--grey-shade-3);
    --active-background: var(--grey-shade-2);
    --enabled-thumb-color: var(--grey-shade-2);
    --tooltip-opacity: 0;
  }

  input:enabled:focus {
    --enabled-thumb-color: var(--primary-color);
  }

  input[variant="success"]:enabled:not(:focus) {
    --active-background: var(--success-color);
  }

  input[variant="success"]:enabled:not(:focus)::-webkit-slider-thumb {
    --enabled-thumb-color: var(--success-color);
  }

  input[variant="warning"]:enabled:not(:focus) {
    --active-background: var(--warning-color);
  }

  input[variant="warning"]:enabled:not(:focus)::-webkit-slider-thumb {
    --enabled-thumb-color: var(--warning-color);
  }

  input[variant="error"]:enabled:not(:focus) {
    --active-background: var(--error-color);
  }

  input[variant="error"]:enabled:not(:focus)::-webkit-slider-thumb {
    --enabled-thumb-color: var(--error-color);
  }

  input {
    appearance: none;
    display: inline-block;
    height: 0.25rem;
    background: #ccc;
    border-radius: 0.125rem;
    margin: 0 var(--component-padding);
  }

  input {
    transition: background-color ease-in-out var(--transition-base),
      border-color ease-in-out var(--transition-base),
      outline ease-in-out var(--transition-base);
    outline-offset: var(--focus-outline-offset);
    outline: var(--focus-outline-inactive);

    border-radius: 16px;
  }

  label {
    transition: color ease-in-out var(--transition-base);
  }

  input::-webkit-slider-thumb {
    -webkit-appearance: none;
    appearance: none;
    transition: background-color ease-in-out var(--transition-base);
  }

  input::-moz-range-thumb {
    width: 1rem;
    height: 1rem;
    border-radius: 50%;
    background: var(--enabled-thumb-color);
    cursor: pointer;
  }

  input::-webkit-slider-thumb {
    width: 1rem;
    height: 1rem;
    border-radius: 50%;
    background: var(--enabled-thumb-color);
    cursor: pointer;
  }

  input[readonly=""]:enabled:focus-visible::-webkit-slider-thumb {
    background: var(--inactive-background);
    cursor: default;
    pointer-events: none;
  }

  input[readonly=""]:enabled:focus::-webkit-slider-thumb {
    background: var(--enabled-thumb-color);
    cursor: default;
    pointer-events: none;
  }

  input:disabled {
    --active-background: var(--inactive-background);
  }

  input:disabled::-webkit-slider-thumb {
    background: var(--inactive-background);
  }

  input:enabled:focus,
  input:enabled:focus-visible {
    --active-background: var(--primary-color);
    outline-offset: var(--focus-outline-offset);
    outline: var(--focus-outline);
    --tooltip-opacity: 1;
  }

  label {
    margin-bottom: 0.5rem;
    text-transform: uppercase;
  }

  label:has(+ div > div > input:disabled) {
    color: var(--text-color-disabled);
  }

  label:has(+ div > div > input:enabled:focus) {
    color: var(--primary-color);
  }

  label:has(+ div > div > input[variant="success"]:not(:focus)) {
    color: var(--success-color);
  }
  label:has(+ div > div > input[variant="warning"]:not(:focus)) {
    color: var(--warning-color);
  }

  label:has(+ div > div > input[variant="error"]:not(:focus)) {
    color: var(--error-color);
  }

  #message {
    transition: max-height ease-in-out var(--transition-base);
  }

  #message {
    max-height: unset;
    font-size: 0.75rem;
    margin: 0.5rem 0 0.5rem 0.125rem;
  }

  div:has(input[variant="warning"]) + #message {
    color: var(--warning-color);
  }

  div:has(input[variant="success"]) + #message {
    color: var(--success-color);
  }

  div:has(input[variant="error"]) + #message {
    color: var(--error-color);
  }

  .layer:has(input:focus) + .tooltip-aligner {
    --tooltip-opacity: 1;
  }

  #message:empty {
    max-height: 0;
    display: none;
  }

  :host {
    display: inline-flex;
    flex-direction: column;
    padding: var(--component-padding);
  }

  .container {
    position: relative;
    height: 0.75rem;
    display: flex;
  }

  .layer {
    display: flex;
    align-items: center;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
  }

  #tooltip {
    transition: opacity ease-in-out var(--transition-base);
  }

  #tooltip {
    opacity: var(--tooltip-opacity);
    position: absolute;
    display: inline-block;
    padding: 0.25rem 0.75rem;
    border-radius: 1.5rem;
    min-width: min-content;

    max-width: min-content;
    top: -2rem;
    left: 0;
    right: 0;
    transform: translateX(-50%);
    color: var(--background-color);
    background-color: var(--grey-shade-2);
  }

  .tooltip-aligner {
    position: absolute;
    margin-left: 2.75rem;
    margin-right: 2.75rem;
    width: calc(100% - 5.5rem);
  }

  div > input {
    flex: 1;
  }

  span {
    color: var(--text-color);
    min-width: 2rem;
    font-size: 0.75rem;
  }
</style>
<label></label>
<div class="container">
  <div class="layer">
    <span>0%</span>
    <input type="range" min="0" max="100" value="50" style="">
    <span>100%</span>
  </div>
  <div class="tooltip-aligner">
    <div id="tooltip" style="left: 50%;">50</div>
  </div>
</div>
<p id="message"></p>
"
`;
